### ISSUE TITLE
Slow batch inserts into Buffer table

### ISSUE THREAD

**Issue by valyala** [yandex/ClickHouse#594]
## Steps to reproduce

* Create a `Buffer` table with multiple buffers (16 in this case) for the existing `MergeTree` table:

```sql
CREATE TABLE foo_buffer AS foo ENGINE = Buffer(default, foo, 16, 10, 60, 100000, 20000000, 10000000, 200000000);
```

* Compare insert times for identical batches between `foo_buffer` and `foo` tables.

  Note that the majority of insert times for `foo_buffer` are smaller than the insert times for `foo` table, but a small percent of the insert times for `foo_buffer` significantly exceeds insert times for `foo`.

## Thoughts

It looks like the `clickhouse` blocks the current insert into `foo_buffer` if the currently chosen buffer is flushed into `foo`. It would be better to try inserting into another buffers instead of waiting until the initially chosen buffer is flushed.

**Comment by ludv1x** [comment]
Yes, such effect is exists.
Index of buffer is chosen by formula `thread_number() % num_buffers`, and buffer is locked under exclusive mutex if other thread use it.

**Comment by ludv1x** [comment]
I described the algorithm in previous comment inaccurately.
It starts from `thread_number() % num_buffers`-th buffer and tries to insert new block into the first unlocked buffer. 

But it doesn't consider free space in a buffer.
I changed behavior, now it tries to insert block into unlocked buffer with max free space.

It slightly improves performance:

20K rows inserts 20M buffers:
![query_time_percentiles_20k_buf20m](https://cloud.githubusercontent.com/assets/1247636/24202327/c93112b4-0f23-11e7-826f-6721e32698ef.png)

20K rows inserts 200K buffers:
![query_time_percentiles_20k_buf200k](https://cloud.githubusercontent.com/assets/1247636/24202338/cf9b00ba-0f23-11e7-8e07-c38fddd180db.png)
